force_wfmanager/utils/variable_names_registry.py

```diff
@@ -2,221 +2,326 @@
     HasStrictTraits, List, Instance, on_trait_change, Property,
-    cached_property, Dict, Tuple)
+    cached_property, Dict, Tuple, HasTraits, Int, Unicode)
 
-from force_bdss.api import Identifier, Workflow
+from force_bdss.api import (
+    Identifier, Workflow, BaseDataSourceModel, OutputSlotInfo,
+    InputSlotInfo, VerifierError
+)
 from force_bdss.local_traits import CUBAType
 
 log = logging.getLogger(__name__)
 
 
-class VariableNamesRegistry(HasStrictTraits):
-    """ Class used for listening to the structure of the Workflow in order to
-    check the available variables that can be used as inputs for each layer
-    """
+class Variable(HasTraits):
+    """Class used to store UI input and output information from DataSources
+    as individual variables. These are then selected by the MCO optimiser as
+    possible parameters or KPIs"""
 
     # -------------------
     # Required Attributes
     # -------------------
 
-    #: Workflow model
-    workflow = Instance(Workflow, allow_none=False)
+    # CUBA type of Variable
+    type = CUBAType()
 
     # ------------------
     # Regular Attributes
     # ------------------
 
-    #: For each execution layer, there will be a list of (name, type) pairs
-    #: representing the input variables produced by that execution layer
-    data_source_input = List(Tuple(Identifier, CUBAType))
+    # Layer index at which variable is generated
+    layer = Int
 
-    #: For each execution layer, there will be a list of (name, type) pairs
-    #: representing the output variables produced by that execution layer
-    data_source_output = List(Tuple(Identifier, CUBAType))
+    # DataSource that contains Variable as an output
+    origin = Instance(BaseDataSourceModel)
 
-    #: For each execution layer, there will be a list of (name, type) pairs
-    #: representing the input variables produced by that execution layer
-    exec_layer_input = List(data_source_input)
+    # DataSource output slot that is used to update Variable name
+    output_slot = Instance(OutputSlotInfo)
 
-    #: For each execution layer, there will be a list of (name, type) pairs
-    #: representing the output variables produced by that execution layer
-    exec_layer_output = List(data_source_output)
+    # DataSources where Variable is used as an input
+    input_slots = List(Tuple(Int, InputSlotInfo))
 
-    #: Dictionary allowing the lookup of names associated with a chosen CUBA
-    #: type.
-    exec_layer_by_type = Dict(CUBAType, List(Identifier))
+    # ---------------------
+    # Dependent Attributes
+    # ---------------------
+
+    # Name of Variable, listens to output_slot if updated by UI
+    name = Identifier()
+
+    # ------------------
+    #     Properties
+    # ------------------
+
+    label = Property(Unicode, depends_on='output_slot.name,type')
+
+    # ------------------
+    #     Listeners
+    # ------------------
+
+    def _get_label(self):
+        return f'{self.type} {self.name}'
```

Comment:
Discussed offline: suggest using a tuple (type, name) to avoid ambiguity

Reply:
Included in latest commit